src/hdmf_zarr/backend.py

```diff
@@ -546,14 +547,41 @@ def write_group(self, **kwargs):
                     export_source=export_source,
                 )
 
-        # write all links (haven implemented)
         links = builder.links
         if links:
+            # Note: In HDMF, we simply call write_link (within write_group) which contains a similar logic as seen
+            # below. This is not the case here (until a refactor).
             for link_name, sub_builder in links.items():
-                self.write_link(group, sub_builder)
+                group_filename = self.__get_store_path(group.store)
+                if export_source is not None:
+                    if sub_builder.builder.source in (group_filename, export_source):
+                        # Case 1:
+                        # sub_builder.builder.source == export_source
+                        # This means we have a SoftLink for a group and so we want the exported link to
+                        # also point "inwards" in the file being created.
+                        #################################
+                        # Case 2:
+                        # sub_builder.builder.source == group_filename
+                        # This is still a SoftLink; however, it is from adding a link to a group after FileA
+                        # has been read and we are exporting that to FileB. We still want the link to be "inwards".
+                        source = group_filename
```

Comment:
It's confusing that the variable is called `source` but then it is used to set `export_source` on the `write_link` function. Moving the logic to `write_link` will probably help, be we should name the variable according to what it is used for, since this is not the actual "source" you are using.